pallets/foreign-investments/src/impls/redeem.rs

```diff
@@ -584,3 +584,158 @@ where
+				Self::RedeemingAndSwapIntoForeignDone { done_swap, .. } => done_swap.amount,
+				_ => 0,
+			}
+		}
+
+		fn get_swap_amount(&self) -> u128 {
+			match *self {
+				Self::ActiveSwapIntoForeignCurrency { swap } => swap.amount,
+				Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone { swap, .. } => {
+					swap.amount
+				}
+				Self::RedeemingAndActiveSwapIntoForeignCurrency { swap, .. } => swap.amount,
+				Self::RedeemingAndActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+					swap,
+					..
+				} => swap.amount,
+				_ => 0,
+			}
+		}
+
+		fn total(&self) -> u128 {
+			self.get_redeeming_amount() + self.get_done_amount() + self.get_swap_amount()
+		}
+	}
+
+	struct Checker {
+		old_state: RedeemState,
+	}
+
+	impl Checker {
+		fn new(initial_state: RedeemState, use_case: &[RedeemTransition]) -> Self {
+			println!("Testing use case: {:#?}", use_case);
+
+			Self {
+				old_state: initial_state,
+			}
+		}
+
+		/// Invariants from: https://centrifuge.hackmd.io/IPtRlOrOSrOF9MHjEY48BA?view#Without-storage
+		fn check_delta_invariant(&self, transition: &RedeemTransition, new_state: &RedeemState) {
+			println!("Transition: {:#?}", transition);
+			println!("New state: {:#?}", new_state);
+
+			match *transition {
+				RedeemTransition::IncreaseRedeemOrder(amount) => {
+					let diff = new_state.total() - self.old_state.total();
+					assert_eq!(diff, amount);
+				}
+				RedeemTransition::DecreaseRedeemOrder(amount) => {
+					let diff = self.old_state.total() - new_state.total();
+					assert_eq!(diff, amount);
+				}
+				RedeemTransition::FulfillSwapOrder(swap) => {
+					let diff = new_state.total() - self.old_state.total();
+					assert_eq!(diff, 0);
+
+					let done_diff = new_state.get_done_amount() - self.old_state.get_done_amount();
+					assert_eq!(done_diff, swap.amount)
+				}
+				RedeemTransition::CollectRedemption(value, swap) => {
+					if self.old_state.get_redeeming_amount() == 0 {
+						assert_eq!(new_state.get_redeeming_amount(), 0)
+					} else {
+						assert_eq!(new_state.get_redeeming_amount(), value);
+					}
+
+					let swap_diff = new_state.get_swap_amount() - self.old_state.get_swap_amount();
+					assert_eq!(swap_diff, swap.amount)
+				}
+			};
+		}
+	}
+
+	#[test]
+	fn fuzzer() {
+		let foreign_swap_big = Swap {
+			currency_in: CurrencyId::Foreign,
+			currency_out: CurrencyId::Pool,
+			amount: 120,
+		};
+		let foreign_swap_small = Swap {
+			currency_in: CurrencyId::Foreign,
+			currency_out: CurrencyId::Pool,
+			amount: 60,
+		};
+
+		let transitions = [
+			RedeemTransition::IncreaseRedeemOrder(120),
+			RedeemTransition::IncreaseRedeemOrder(60),
+			RedeemTransition::DecreaseRedeemOrder(120),
+			RedeemTransition::DecreaseRedeemOrder(60),
+			RedeemTransition::FulfillSwapOrder(foreign_swap_big),
+			RedeemTransition::FulfillSwapOrder(foreign_swap_small),
+			//RedeemTransition::CollectInvestment(60),
+			//RedeemTransition::CollectInvestment(120),
```

Comment:
I suppose we also want to check `CollectRedemption`.

Reply:
Right! I didn't know what I did there to have them commented 😅. 
Updated and working!